core/src/main/scala/cats/Functor.scala

```diff
@@ -177,6 +177,19 @@ trait Functor[F[_]] extends Invariant[F] { self =>
    */
   def tupleRight[A, B](fa: F[A], b: B): F[(A, B)] = map(fa)(a => (a, b))
 
+  /**
+   * Modifies the `A` value in `F[A]` with the supplied function, if the function is defined for the value.
+   * Example:
+   * {{{
+   * scala> import cats.Functor
+   * scala> import cats.implicits.catsStdInstancesForList
+   *
+   * scala> Functor[List].alter(List(1, 2, 3)) { case 2 => 42 }
+   * res0: List[Int] = List(1, 42, 3)
+   * }}}
+   */
+  def alter[A, A1 >: A](fa: F[A])(pf: PartialFunction[A, A1]): F[A1] = map(fa)(a => if (pf.isDefinedAt(a)) pf(a) else a)
```

Comment:
We should add a law to FunctorLaws that all implementations match version.

A possible compiling but wrong version would be to return `fa.widen[A1]`

Reply:
Calling to `pf` twice - `isDefinedAt` and then `apply` is not very efficient. Usually `applyOrElse` should be used where possible:
```suggestion
  def alter[A, A1 >: A](fa: F[A])(pf: PartialFunction[A, A1]): F[A1] =
    map(fa)(a => pf.applyOrElse(a, _ => a))
```

Reply:
from `applyOrElse` doc
```
Note that expression pf.applyOrElse(x, default) is equivalent to
if(pf isDefinedAt x) pf(x) else default(x) 
```
and here's the implementation
```scala
  def applyOrElse[A1 <: A, B1 >: B](x: A1, default: A1 => B1): B1 =
    if (isDefinedAt(x)) apply(x) else default(x)
```
how is this more efficient?

Reply:
Thanks for the suggestion, Oscar. 
do you mean something like
```scala
  def functorAlterIdentity[A, A1 >: A](fa: F[A]): IsEq[F[A1]] =
    fa.alter { case a => a: A1 } <-> fa.widen[A1]

  def functorAlter[A, A1 >: A](fa: F[A], pf: PartialFunction[A, A1]): IsEq[F[A1]] =
    fa.alter(pf) <-> fa.map(a => if (pf.isDefinedAt(a)) pf(a) else a)
```
?

Reply:
@jozic ,
> from applyOrElse doc ..

correct, but also the doc says that:
>   *  Note that expression `pf.applyOrElse(x, default)` is equivalent to
>   *  {{{ if(pf isDefinedAt x) pf(x) else default(x) }}}
>   *  except that `applyOrElse` method can be implemented more efficiently.

In other words, when `PartialFunction` is auto-generated by the compiler (or composed via `orElse` for example), its `applyOrElse` can be other than the default implementation.

Reply:
makes sense, i didn't read the doc fully, my bad
will update to `applyOrElse`

Reply:
Naming is notoriously known for being one of two hard things in CS. AFAIU, `alter` isn't widespread in Haskell's standard library (reference implementation ftw), so maybe we can reflect a bit on naming to make it more intuitive from the code-completion standpoint.
We already have some variations of `map` — `mapF`, `mapK`, `map2`, `mapFilter`, `mapWithIndex` and maybe something else. Given that, perhaps, `mapPartially` would pursue coherence in naming when dealing with some kind of `map` in cats and be slightly more versatile from a code-completion applicability.
WDYT?

Reply:
I'm not able to search now, but we have a couple of other places where we take partial functions. We be good to review the choices there. Here are some possible candidates:

- mapWhen
- mapWhere
- mapOrCurrent
- mapOrId
- mapOrRetain
- mapOrKeep

One challenge is that we don't want a reader to confuse it with collect which also takes a PartialFunction, or worse, confuse a writer who intends collect but calls this method, or the opposite.

Reply:
My c2 on suggested options
- `mapWhen` — whenever I see `when` I'd expect a Boolean condition in the contract, i.e. it confuses with `whenA` / `whenF`. But in our case, it's just a `PartialFunction`. 
- `mapWhere` — we don't yet have a `where` prefix/postfix in API. Although it might sound natural in pure English `... map Where string length > 0`, I can't say I like this option.

Reply:
From the options above I personally like `mapOrKeep` because it is concise but still quite self-descriptive.

Reply:
Agreed, `mapOrKeep` works well for me. But what about `mapPartially`? It's a straightforward one. Although, not arguing to settle on it, just wondering about arguments against it.

Reply:
It is more like the matter of flavors of English language (which I am not a great expert in 😄 so I may be wrong), but to me personally, `mapPartially` may sound more like "map-or-skip" rather than "map-or-keep". The former is going to be `collect` then. In other words, `mapPartially` looks a bit less definitive. Not that critical though – both would work for me too.

Reply:
I've checked usages of PF in this repo and I've found following methods
- `collect`/`collectFirst`/`collectFold`/`traverseCollect`
- `adaptError`/`onError`/`recover`/`recoverWith`
- `reject` in `monadError`

I personally like `mapWhen` and i get a link to boolean here, but PF seems like a function with a predicate (which is sort of boolean), `list.mapWhen { case i if i > 1 => 1 }`. It reads intuitively (subjectively, of course) as `map when condition in  case statements are true`.
But at the end of the day i'm fine with `mapOrKeep`, so i used that name for now.
Please let me know if someone shares my `mapWhen` sentiment